# Exploratory Data Analysis

In [18]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [19]:
file_path = "./DataSets/"
training_data = pd.read_csv(file_path + 'counterfeit_train.csv')
production_data = pd.read_csv(file_path + 'counterfeit_test.csv')

In [20]:
training_data.head()

Medicine_ID  Counterfeit_Weight DistArea_ID  Active_Since  ...           Area_Type Area_City_Type Area_dist_level  Counterfeit_Sales
0       RRA15              13.100     Area046          1995  ...            DownTown         Tier 1           Small          1775.5026
1       YVV26                 NaN     Area027          1983  ...          CityLimits         Tier 3          Medium          3069.1520
2       LJC15               9.025     Area046          1995  ...            DownTown         Tier 1           Small          2603.0920
3       GWC40              11.800     Area046          1995  ...            DownTown         Tier 1           Small          1101.7130
4       QMN13                 NaN     Area019          1983  ...  MidTownResidential         Tier 1           Small           158.9402

[5 rows x 12 columns]

In [21]:
production_data.head()

Medicine_ID  Counterfeit_Weight DistArea_ID  Active_Since  ...  Availability_rating           Area_Type Area_City_Type  Area_dist_level
0       HLZ81                 NaN     Area027          1983  ...             0.112747          CityLimits         Tier 3           Medium
1       ECE94               13.45     Area045          2000  ...             0.144446            DownTown         Tier 2          Unknown
2       SAD14                7.10     Area045          2000  ...             0.144221            DownTown         Tier 2          Unknown
3       EQV63               18.30     Area010          1996  ...             0.100388  MidTownResidential         Tier 3          Unknown
4       AIR10                 NaN     Area019          1983  ...             0.022585  MidTownResidential         Tier 1            Small

[5 rows x 11 columns]

In [22]:
training_data['Counterfeit_Weight'].mean()

14.115056617126996

In [23]:
training_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6818 entries, 0 to 6817
Data columns (total 12 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Medicine_ID          6818 non-null   object 
 1   Counterfeit_Weight   5652 non-null   float64
 2   DistArea_ID          6818 non-null   object 
 3   Active_Since         6818 non-null   int64  
 4   Medicine_MRP         6818 non-null   float64
 5   Medicine_Type        6818 non-null   object 
 6   SidEffect_Level      6818 non-null   object 
 7   Availability_rating  6818 non-null   float64
 8   Area_Type            6818 non-null   object 
 9   Area_City_Type       6818 non-null   object 
 10  Area_dist_level      6818 non-null   object 
 11  Counterfeit_Sales    6818 non-null   float64
dtypes: float64(4), int64(1), object(7)
memory usage: 639.3+ KB


In [24]:
production_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1705 entries, 0 to 1704
Data columns (total 11 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Medicine_ID          1705 non-null   object 
 1   Counterfeit_Weight   1408 non-null   float64
 2   DistArea_ID          1705 non-null   object 
 3   Active_Since         1705 non-null   int64  
 4   Medicine_MRP         1705 non-null   float64
 5   Medicine_Type        1705 non-null   object 
 6   SidEffect_Level      1705 non-null   object 
 7   Availability_rating  1705 non-null   float64
 8   Area_Type            1705 non-null   object 
 9   Area_City_Type       1705 non-null   object 
 10  Area_dist_level      1705 non-null   object 
dtypes: float64(3), int64(1), object(7)
memory usage: 146.6+ KB


Creating a Dummy sales variable in the production data and combining it with training data for further data manipulations

In [25]:
production_data['Counterfeit_Sales'] = np.NAN

In [26]:
training_data['Data_from'] = 'Train'
production_data['Data_from'] = 'Production'

In [27]:
full_data = pd.concat([training_data,production_data],axis=0)

In [28]:
full_data.loc[full_data['Data_from'] == 'Train',]

Medicine_ID  Counterfeit_Weight DistArea_ID  Active_Since  ...  Area_City_Type Area_dist_level Counterfeit_Sales  Data_from
0          RRA15              13.100     Area046          1995  ...          Tier 1           Small         1775.5026      Train
1          YVV26                 NaN     Area027          1983  ...          Tier 3          Medium         3069.1520      Train
2          LJC15               9.025     Area046          1995  ...          Tier 1           Small         2603.0920      Train
3          GWC40              11.800     Area046          1995  ...          Tier 1           Small         1101.7130      Train
4          QMN13                 NaN     Area019          1983  ...          Tier 1           Small          158.9402      Train
...          ...                 ...         ...           ...  ...             ...             ...               ...        ...
6813       OYN80               8.535     Area046          1995  ...          Tier 1           Small         2070.4520      Train
6814       ACW12              20.650     Area046          1995  ...          Tier 1           Small         2126.3792      Train
6815       OPM10              20.000     Area017          2005  ...          Tier 2         Unknown         2119.7212      Train
6816       SLY12              10.180     Area045          2000  ...          Tier 2         Unknown         1485.2138      Train
6817       ATT10               7.930     Area018          2007  ...          Tier 3          Medium         1943.9500      Train

[6818 rows x 13 columns]

In [29]:
full_data['DistArea_ID'].value_counts()

Area027    935
Area013    932
Area046    930
Area049    930
Area035    930
Area045    929
Area018    928
Area017    926
Area010    555
Area019    528
Name: DistArea_ID, dtype: int64

In [30]:
full_data['Area_Type'].value_counts()

DownTown              5577
MidTownResidential    1083
CityLimits             935
Industrial             928
Name: Area_Type, dtype: int64

In [31]:
full_data.loc[full_data['Area_Type']=='DownTown','DistArea_ID'].unique()

array(['Area046', 'Area045', 'Area013', 'Area017', 'Area035', 'Area049'],
      dtype=object)

In [32]:
full_data.loc[full_data['Area_Type']=='MidTownResidential','DistArea_ID'].unique()

array(['Area019', 'Area010'], dtype=object)

In [33]:
full_data.loc[full_data['Area_Type']=='CityLimits','DistArea_ID'].unique()

array(['Area027'], dtype=object)

In [34]:
full_data.loc[full_data['Area_Type']=='Industrial','DistArea_ID'].unique()

array(['Area018'], dtype=object)

In [35]:
full_data.head()

Medicine_ID  Counterfeit_Weight DistArea_ID  Active_Since  ...  Area_City_Type Area_dist_level Counterfeit_Sales  Data_from
0       RRA15              13.100     Area046          1995  ...          Tier 1           Small         1775.5026      Train
1       YVV26                 NaN     Area027          1983  ...          Tier 3          Medium         3069.1520      Train
2       LJC15               9.025     Area046          1995  ...          Tier 1           Small         2603.0920      Train
3       GWC40              11.800     Area046          1995  ...          Tier 1           Small         1101.7130      Train
4       QMN13                 NaN     Area019          1983  ...          Tier 1           Small          158.9402      Train

[5 rows x 13 columns]

In [36]:
full_data['Medicine_Type'].value_counts()

Antibiotics           1232
Hreplacements         1200
Antiseptics            910
OralContraceptives     856
Antipyretics           682
Cardiac                649
Mstablizers            648
Tranquilizers          520
Analgesics             445
Antimalarial           425
Antacids               251
Statins                214
MuscleRelaxants        169
Antifungal             148
Stimulants             110
Antiviral               64
Name: Medicine_Type, dtype: int64

In [37]:
round(full_data.groupby('Medicine_Type')['Counterfeit_Sales'].mean(),2)

Medicine_Type
Analgesics            2091.24
Antacids              2259.88
Antibiotics           2397.20
Antifungal            2444.06
Antimalarial          2192.68
Antipyretics          2292.72
Antiseptics           2366.59
Antiviral             2583.16
Cardiac               2362.98
Hreplacements         2376.12
Mstablizers           2044.87
MuscleRelaxants       2091.09
OralContraceptives    2245.22
Statins               2309.45
Stimulants            2279.97
Tranquilizers         2084.45
Name: Counterfeit_Sales, dtype: float64

In [38]:
full_data['SidEffect_Level'].value_counts()

mild        5517
critical    3006
Name: SidEffect_Level, dtype: int64

In [39]:
full_data['Area_dist_level'].value_counts()

Medium     2793
Unknown    2410
Small      2388
High        932
Name: Area_dist_level, dtype: int64

#### After looking into all the categorical  variables, we cannot drop them and we create dummies for each variable for simplicity later we can ckeck for additional options to improve the performance

In [40]:
full_data.head(10)

Medicine_ID  Counterfeit_Weight DistArea_ID  Active_Since  ...  Area_City_Type Area_dist_level Counterfeit_Sales  Data_from
0       RRA15              13.100     Area046          1995  ...          Tier 1           Small         1775.5026      Train
1       YVV26                 NaN     Area027          1983  ...          Tier 3          Medium         3069.1520      Train
2       LJC15               9.025     Area046          1995  ...          Tier 1           Small         2603.0920      Train
3       GWC40              11.800     Area046          1995  ...          Tier 1           Small         1101.7130      Train
4       QMN13                 NaN     Area019          1983  ...          Tier 1           Small          158.9402      Train
5       JDG81               8.775     Area045          2000  ...          Tier 2         Unknown         3047.8464      Train
6       KPX48              18.000     Area018          2007  ...          Tier 3          Medium         2883.3938      Train
7       CYW14               7.680     Area013          1985  ...          Tier 3            High         3262.2340      Train
8       TVF57              18.550     Area046          1995  ...          Tier 1           Small         1959.9292      Train
9       MGU75               9.510     Area013          1985  ...          Tier 3            High         3095.7840      Train

[10 rows x 13 columns]

In [41]:
d = pd.get_dummies(full_data['DistArea_ID'],prefix='DistID',drop_first=True)

In [42]:
full_data = pd.concat((full_data,d),axis='columns')

In [43]:
full_data.drop('DistArea_ID',axis='columns',inplace=True)

In [44]:
full_data

Medicine_ID  Counterfeit_Weight  Active_Since  Medicine_MRP  ... DistID_Area035 DistID_Area045  DistID_Area046 DistID_Area049
0          RRA15              13.100          1995      160.2366  ...              0              0               1              0
1          YVV26                 NaN          1983      110.4384  ...              0              0               0              0
2          LJC15               9.025          1995      259.4092  ...              0              0               1              0
3          GWC40              11.800          1995       99.9830  ...              0              0               1              0
4          QMN13                 NaN          1983       56.4402  ...              0              0               0              0
...          ...                 ...           ...           ...  ...            ...            ...             ...            ...
1700       KXW10                 NaN          1983      136.5704  ...              0              0               0              0
1701       CKE54              21.300          2002       57.0744  ...              1              0               0              0
1702       HAY13              20.400          2005      182.7422  ...              0              0               0              0
1703       ZEE32              20.000          2007      266.9672  ...              0              0               0              0
1704       NDL93               8.080          2000      238.7694  ...              0              1               0              0

[8523 rows x 21 columns]

##### Medicine variable - Dummy Creations

In [45]:
d = pd.get_dummies(full_data['Medicine_Type'],prefix='MedType',drop_first=True)
full_data = pd.concat((full_data,d),axis='columns')
full_data.drop('Medicine_Type',axis='columns',inplace=True,)

In [46]:
full_data

Medicine_ID  Counterfeit_Weight  Active_Since  ...  MedType_Statins MedType_Stimulants  MedType_Tranquilizers
0          RRA15              13.100          1995  ...                0                  0                      0
1          YVV26                 NaN          1983  ...                0                  0                      0
2          LJC15               9.025          1995  ...                0                  0                      0
3          GWC40              11.800          1995  ...                0                  0                      0
4          QMN13                 NaN          1983  ...                0                  0                      0
...          ...                 ...           ...  ...              ...                ...                    ...
1700       KXW10                 NaN          1983  ...                0                  0                      0
1701       CKE54              21.300          2002  ...                0                  0                      0
1702       HAY13              20.400          2005  ...                0                  0                      0
1703       ZEE32              20.000          2007  ...                0                  0                      0
1704       NDL93               8.080          2000  ...                0                  0                      0

[8523 rows x 35 columns]

##### SidEffect_Level - Dummy Creation

In [47]:
full_data['SidEffect_Level'].value_counts()

mild        5517
critical    3006
Name: SidEffect_Level, dtype: int64

In [48]:
d = pd.get_dummies(full_data['SidEffect_Level'],prefix='SidLvl',drop_first=True)
full_data = pd.concat((full_data,d),axis='columns')
full_data.drop('SidEffect_Level',axis='columns',inplace = True)

In [49]:
full_data

Medicine_ID  Counterfeit_Weight  Active_Since  Medicine_MRP  ...  MedType_Statins MedType_Stimulants MedType_Tranquilizers SidLvl_mild
0          RRA15              13.100          1995      160.2366  ...                0                  0                     0           0
1          YVV26                 NaN          1983      110.4384  ...                0                  0                     0           1
2          LJC15               9.025          1995      259.4092  ...                0                  0                     0           1
3          GWC40              11.800          1995       99.9830  ...                0                  0                     0           1
4          QMN13                 NaN          1983       56.4402  ...                0                  0                     0           0
...          ...                 ...           ...           ...  ...              ...                ...                   ...         ...
1700       KXW10                 NaN          1983      136.5704  ...                0                  0                     0           1
1701       CKE54              21.300          2002       57.0744  ...                0                  0                     0           0
1702       HAY13              20.400          2005      182.7422  ...                0                  0                     0           1
1703       ZEE32              20.000          2007      266.9672  ...                0                  0                     0           1
1704       NDL93               8.080          2000      238.7694  ...                0                  0                     0           0

[8523 rows x 35 columns]

##### Area_Type - Dummy Creation

In [50]:
d = pd.get_dummies(full_data['Area_Type'],prefix='ArTyp',drop_first=True)
full_data = pd.concat((full_data,d),axis='columns')
full_data.drop('Area_Type',axis='columns',inplace = True)

In [51]:
full_data

Medicine_ID  Counterfeit_Weight  Active_Since  Medicine_MRP  ...  SidLvl_mild ArTyp_DownTown ArTyp_Industrial  ArTyp_MidTownResidential
0          RRA15              13.100          1995      160.2366  ...            0              1                0                         0
1          YVV26                 NaN          1983      110.4384  ...            1              0                0                         0
2          LJC15               9.025          1995      259.4092  ...            1              1                0                         0
3          GWC40              11.800          1995       99.9830  ...            1              1                0                         0
4          QMN13                 NaN          1983       56.4402  ...            0              0                0                         1
...          ...                 ...           ...           ...  ...          ...            ...              ...                       ...
1700       KXW10                 NaN          1983      136.5704  ...            1              0                0                         0
1701       CKE54              21.300          2002       57.0744  ...            0              1                0                         0
1702       HAY13              20.400          2005      182.7422  ...            1              1                0                         0
1703       ZEE32              20.000          2007      266.9672  ...            1              0                1                         0
1704       NDL93               8.080          2000      238.7694  ...            0              1                0                         0

[8523 rows x 37 columns]

##### Area_dist_level - Dummy Creation

In [52]:
d = pd.get_dummies(full_data['Area_dist_level'],prefix='ArDisTyp',drop_first=True)
full_data = pd.concat((full_data,d),axis='columns')
full_data.drop('Area_dist_level',axis='columns',inplace = True)

In [53]:
full_data

Medicine_ID  Counterfeit_Weight  Active_Since  Medicine_MRP  ...  ArTyp_MidTownResidential ArDisTyp_Medium  ArDisTyp_Small ArDisTyp_Unknown
0          RRA15              13.100          1995      160.2366  ...                         0               0               1                0
1          YVV26                 NaN          1983      110.4384  ...                         0               1               0                0
2          LJC15               9.025          1995      259.4092  ...                         0               0               1                0
3          GWC40              11.800          1995       99.9830  ...                         0               0               1                0
4          QMN13                 NaN          1983       56.4402  ...                         1               0               1                0
...          ...                 ...           ...           ...  ...                       ...             ...             ...              ...
1700       KXW10                 NaN          1983      136.5704  ...                         0               1               0                0
1701       CKE54              21.300          2002       57.0744  ...                         0               0               1                0
1702       HAY13              20.400          2005      182.7422  ...                         0               0               0                1
1703       ZEE32              20.000          2007      266.9672  ...                         0               1               0                0
1704       NDL93               8.080          2000      238.7694  ...                         0               0               0                1

[8523 rows x 39 columns]

##### Area_City_Type 	- Dummy Creation

In [54]:
d = pd.get_dummies(full_data['Area_City_Type'],prefix='ArCtyTyp',drop_first=True)
full_data = pd.concat((full_data,d),axis='columns')
full_data.drop('Area_City_Type',axis='columns',inplace = True)

In [55]:
full_data

Medicine_ID  Counterfeit_Weight  Active_Since  Medicine_MRP  ...  ArDisTyp_Small  ArDisTyp_Unknown ArCtyTyp_Tier 2  ArCtyTyp_Tier 3
0          RRA15              13.100          1995      160.2366  ...               1                 0               0                0
1          YVV26                 NaN          1983      110.4384  ...               0                 0               0                1
2          LJC15               9.025          1995      259.4092  ...               1                 0               0                0
3          GWC40              11.800          1995       99.9830  ...               1                 0               0                0
4          QMN13                 NaN          1983       56.4402  ...               1                 0               0                0
...          ...                 ...           ...           ...  ...             ...               ...             ...              ...
1700       KXW10                 NaN          1983      136.5704  ...               0                 0               0                1
1701       CKE54              21.300          2002       57.0744  ...               1                 0               1                0
1702       HAY13              20.400          2005      182.7422  ...               0                 1               1                0
1703       ZEE32              20.000          2007      266.9672  ...               0                 0               0                1
1704       NDL93               8.080          2000      238.7694  ...               0                 1               1                0

[8523 rows x 40 columns]

##### Dropping Medicine ID

In [56]:
full_data.drop('Medicine_ID',axis='columns',inplace=True)

In [57]:
full_data

Counterfeit_Weight  Active_Since  Medicine_MRP  Availability_rating  ...  ArDisTyp_Small ArDisTyp_Unknown  ArCtyTyp_Tier 2  ArCtyTyp_Tier 3
0                 13.100          1995      160.2366             0.070422  ...               1                0                0                0
1                    NaN          1983      110.4384             0.013000  ...               0                0                0                1
2                  9.025          1995      259.4092             0.060783  ...               1                0                0                0
3                 11.800          1995       99.9830             0.065555  ...               1                0                0                0
4                    NaN          1983       56.4402             0.248859  ...               1                0                0                0
...                  ...           ...           ...                  ...  ...             ...              ...              ...              ...
1700                 NaN          1983      136.5704             0.050505  ...               0                0                0                1
1701              21.300          2002       57.0744             0.041118  ...               1                0                1                0
1702              20.400          2005      182.7422             0.191273  ...               0                1                1                0
1703              20.000          2007      266.9672             0.013000  ...               0                0                0                1
1704               8.080          2000      238.7694             0.098982  ...               0                1                1                0

[8523 rows x 39 columns]

##### Imputing Counterfeit_weight missing values with mean from train dataset only

In [58]:
wgt_mean = round(full_data.loc[full_data['Data_from'] == 'Train','Counterfeit_Weight'].mean(),3)


In [59]:
wgt_mean

14.115

In [60]:
full_data['Counterfeit_Weight'].fillna(wgt_mean,inplace=True)

In [61]:
full_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8523 entries, 0 to 1704
Data columns (total 39 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Counterfeit_Weight          8523 non-null   float64
 1   Active_Since                8523 non-null   int64  
 2   Medicine_MRP                8523 non-null   float64
 3   Availability_rating         8523 non-null   float64
 4   Counterfeit_Sales           6818 non-null   float64
 5   Data_from                   8523 non-null   object 
 6   DistID_Area013              8523 non-null   uint8  
 7   DistID_Area017              8523 non-null   uint8  
 8   DistID_Area018              8523 non-null   uint8  
 9   DistID_Area019              8523 non-null   uint8  
 10  DistID_Area027              8523 non-null   uint8  
 11  DistID_Area035              8523 non-null   uint8  
 12  DistID_Area045              8523 non-null   uint8  
 13  DistID_Area046              8523 

##### Now Saeperating data into Training and Production

In [62]:
training_data = full_data.loc[full_data['Data_from'] == 'Train']
training_data.drop('Data_from',axis=1,inplace=True)

/Users/santu/opt/anaconda3/lib/python3.7/site-packages/pandas/core/frame.py:3997: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


In [63]:
training_data

Counterfeit_Weight  Active_Since  Medicine_MRP  Availability_rating  ...  ArDisTyp_Small  ArDisTyp_Unknown  ArCtyTyp_Tier 2  ArCtyTyp_Tier 3
0                 13.100          1995      160.2366             0.070422  ...               1                 0                0                0
1                 14.115          1983      110.4384             0.013000  ...               0                 0                0                1
2                  9.025          1995      259.4092             0.060783  ...               1                 0                0                0
3                 11.800          1995       99.9830             0.065555  ...               1                 0                0                0
4                 14.115          1983       56.4402             0.248859  ...               1                 0                0                0
...                  ...           ...           ...                  ...  ...             ...               ...              ...              ...
6813               8.535          1995      204.1452             0.112963  ...               1                 0                0                0
6814              20.650          1995      235.1088             0.131103  ...               1                 0                0                0
6815              20.000          2005      193.6292             0.105096  ...               0                 1                1                0
6816              10.180          2000      162.8682             0.099957  ...               0                 1                1                0
6817               7.930          2007      177.5500             0.117445  ...               0                 0                0                1

[6818 rows x 38 columns]

In [64]:
production_data = full_data.loc[full_data['Data_from'] == 'Production']

In [65]:
production_data.drop(['Counterfeit_Sales','Data_from'],axis=True,inplace=True)
production_data

/Users/santu/opt/anaconda3/lib/python3.7/site-packages/pandas/core/frame.py:3997: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


Counterfeit_Weight  Active_Since  Medicine_MRP  Availability_rating  ...  ArDisTyp_Small  ArDisTyp_Unknown  ArCtyTyp_Tier 2  ArCtyTyp_Tier 3
0                 14.115          1983       85.5328             0.112747  ...               0                 0                0                1
1                 13.450          2000      257.1460             0.144446  ...               0                 1                1                0
2                  7.100          2000       98.1172             0.144221  ...               0                 1                1                0
3                 18.300          1996      135.3730             0.100388  ...               0                 1                0                1
4                 14.115          1983      112.8016             0.022585  ...               1                 0                0                0
...                  ...           ...           ...                  ...  ...             ...               ...              ...              ...
1700              14.115          1983      136.5704             0.050505  ...               0                 0                0                1
1701              21.300          2002       57.0744             0.041118  ...               1                 0                1                0
1702              20.400          2005      182.7422             0.191273  ...               0                 1                1                0
1703              20.000          2007      266.9672             0.013000  ...               0                 0                0                1
1704               8.080          2000      238.7694             0.098982  ...               0                 1                1                0

[1705 rows x 37 columns]

# Splitting training data into Train and Test

In [66]:
from sklearn.model_selection import train_test_split

In [67]:
train,test = train_test_split(training_data,test_size=0.2,random_state = 2,)

In [68]:
x_train = train.drop('Counterfeit_Sales',axis='columns')
y_train = train['Counterfeit_Sales']

x_test = test.drop('Counterfeit_Sales',axis='columns')
y_test = test['Counterfeit_Sales']

print('x_train : ',x_train.shape)
print('y_train : ',y_train.shape)

print('x_test : ',x_test.shape)
print('y_test : ',y_test.shape)

x_train :  (5454, 37)
y_train :  (5454,)
x_test :  (1364, 37)
y_test :  (1364,)


# Gradient Boosting

In [69]:
from xgboost.sklearn import XGBRegressor
from sklearn import metrics
from sklearn.model_selection import RandomizedSearchCV

In [70]:
def my_rmse(actual,predicted):
    """Function to calculate Root Mean Square Error"""
    
    return (((actual-predicted)**2).mean())**1/2

In [71]:
x_train.shape

(5454, 37)

In [103]:
params = {
    "n_estimators" : [200,400,600,800,1000,1200,1400,1600,1800],
    'learning_rate' : [0.01,0.05,0.10,0.15,0.20],
    "max_depth" : list(range(3,10)),
    'min_child_weight' : list(range(1,20)),
    'colsample_bytree' : [0.3,0.4,0.5,0.6,0.7,0.8],
    'max_features' : [5,10,15,20,25,30,35,37],
    'subsample' : [0.3,0.4,0.5,0.6,0.7,0.8]
}
iterations = 200

In [121]:
xgb_reg = XGBRegressor(random_state=3,n_jobs=-1,reg_alpha=0,reg_lambda=1.2,verbosity=1)

In [105]:
random_xgb = RandomizedSearchCV(xgb_reg,param_distributions=params,cv=5,n_iter=iterations,
                               random_state=2,scoring='neg_mean_absolute_error',n_jobs=-1)

In [106]:
random_xgb.fit(x_train,y_train)

[09:24:46] WARNING: src/objective/regression_obj.cu:152: reg:linear is now deprecated in favor of reg:squarederror.


RandomizedSearchCV(cv=5, error_score=nan,
                   estimator=XGBRegressor(base_score=0.5, booster='gbtree',
                                          colsample_bylevel=1,
                                          colsample_bynode=1,
                                          colsample_bytree=1, gamma=0,
                                          importance_type='gain',
                                          learning_rate=0.1, max_delta_step=0,
                                          max_depth=3, min_child_weight=1,
                                          missing=None, n_estimators=100,
                                          n_jobs=-1, nthread=None,
                                          objective='reg:linear',
                                          random_state=3, reg_alpha...
                                        'max_depth': [3, 4, 5, 6, 7, 8, 9],
                                        'max_features': [5, 10, 15, 20, 25, 30,
                                

In [94]:
random_xgb.best_estimator_

XGBRegressor(base_score=0.5, booster='gbtree', colsample_bylevel=1,
             colsample_bynode=1, colsample_bytree=0.8, gamma=0,
             importance_type='gain', learning_rate=0.01, max_delta_step=0,
             max_depth=6, max_features=30, min_child_weight=16, missing=None,
             n_estimators=400, n_jobs=-1, nthread=None, objective='reg:linear',
             random_state=3, reg_alpha=0, reg_lambda=1, scale_pos_weight=1,
             seed=None, silent=None, subsample=0.6, verbosity=1)

In [95]:
random_xgb.cv_results_['mean_test_score']

array([ -803.85014469,  -958.12086478,  -856.13914355,  -754.61795614,
       -1008.04560377,  -859.87362113,  -778.76525679,  -891.3303725 ,
        -942.70385702,  -875.93377881,  -971.78915068,  -884.9732203 ,
        -892.78768651,  -765.18096815,  -901.1832135 ,  -851.68743726,
        -808.36567128,  -823.24176436,  -936.59714158,  -895.34732849,
        -861.04896248,  -843.77856107,  -902.99686196,  -847.96928168,
        -800.78770192,  -895.9247677 ,  -982.6332158 ,  -812.77044947,
        -826.99957367,  -800.25643102,  -880.86883244,  -870.32666364,
        -851.27339727,  -908.692031  ,  -845.40478749,  -770.00847654,
        -784.85847409,  -752.53966322,  -863.08942972,  -768.75354905,
        -842.30635456,  -770.17893662,  -777.42800716,  -937.78526101,
        -759.67615224,  -803.0684515 ,  -775.7312745 ,  -853.32353812,
        -868.31832754,  -780.02084677,  -780.78288367,  -772.33534789,
        -822.27880141,  -766.30494811,  -756.32788206,  -784.45504588,
      

In [107]:
random_xgb.best_score_

-750.1077654328067

In [108]:
random_xgb.best_params_

{'subsample': 0.5,
 'n_estimators': 800,
 'min_child_weight': 14,
 'max_features': 5,
 'max_depth': 4,
 'learning_rate': 0.01,
 'colsample_bytree': 0.8}

In [110]:
xgb = random_xgb.best_estimator_

In [111]:
xgb.fit(x_train,y_train)
pred = xgb.predict(x_test)

[09:26:34] WARNING: src/objective/regression_obj.cu:152: reg:linear is now deprecated in favor of reg:squarederror.


In [112]:
print('performance on test data')
print('Mean Absolute Error = ', metrics.median_absolute_error(y_test,pred))
print('Root Mean Square Error = ',my_rmse(y_test,pred))
print("Our model submission performance : ",1-(metrics.median_absolute_error(y_test,pred)/1600))

performance on test data
Mean Absolute Error =  494.82020566406254
Root Mean Square Error =  605384.4703138274
Our model submission performance :  0.690737371459961


In [113]:
print("performance on train data")
pred_train = xgb.predict(x_train)
print('Mean Absolute Error = ', metrics.median_absolute_error(y_train,pred_train))
print('Root Mean Square Error = ',my_rmse(y_train,pred_train))
print("Our model submission performance : ",1-(metrics.median_absolute_error(y_train,pred_train)/1600))

performance on train data
Mean Absolute Error =  473.319269140625
Root Mean Square Error =  493182.28466352366
Our model submission performance :  0.7041754567871094


In [122]:
random_xgb = RandomizedSearchCV(xgb_reg,param_distributions=params,cv=5,n_iter=2000,
                               random_state=2,scoring='neg_mean_absolute_error',n_jobs=-1,verbose=2)

random_xgb.fit(x_train,y_train)

Fitting 5 folds for each of 2000 candidates, totalling 10000 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 12 concurrent workers.
[Parallel(n_jobs=-1)]: Done  17 tasks      | elapsed:   15.7s
[Parallel(n_jobs=-1)]: Done 138 tasks      | elapsed:  2.6min
[Parallel(n_jobs=-1)]: Done 341 tasks      | elapsed:  6.6min
[Parallel(n_jobs=-1)]: Done 624 tasks      | elapsed: 12.5min
[Parallel(n_jobs=-1)]: Done 989 tasks      | elapsed: 19.7min
[Parallel(n_jobs=-1)]: Done 1434 tasks      | elapsed: 27.9min
[Parallel(n_jobs=-1)]: Done 1961 tasks      | elapsed: 38.6min
[Parallel(n_jobs=-1)]: Done 2568 tasks      | elapsed: 50.3min
[Parallel(n_jobs=-1)]: Done 3257 tasks      | elapsed: 64.7min
[Parallel(n_jobs=-1)]: Done 4026 tasks      | elapsed: 79.2min
[Parallel(n_jobs=-1)]: Done 4877 tasks      | elapsed: 94.6min
[Parallel(n_jobs=-1)]: Done 5808 tasks      | elapsed: 112.1min
[Parallel(n_jobs=-1)]: Done 6821 tasks      | elapsed: 131.9min
[Parallel(n_jobs=-1)]: Done 7914 tasks      | elapsed: 152.2min
[Parallel(n_jobs=-1)]: Done 9089 tasks     

[12:45:34] WARNING: src/objective/regression_obj.cu:152: reg:linear is now deprecated in favor of reg:squarederror.


RandomizedSearchCV(cv=5, error_score=nan,
                   estimator=XGBRegressor(base_score=0.5, booster='gbtree',
                                          colsample_bylevel=1,
                                          colsample_bynode=1,
                                          colsample_bytree=1, gamma=0,
                                          importance_type='gain',
                                          learning_rate=0.1, max_delta_step=0,
                                          max_depth=3, min_child_weight=1,
                                          missing=None, n_estimators=100,
                                          n_jobs=-1, nthread=None,
                                          objective='reg:linear',
                                          random_state=3, reg_alpha...
                                        'max_depth': [3, 4, 5, 6, 7, 8, 9],
                                        'max_features': [5, 10, 15, 20, 25, 30,
                                

In [123]:
random_xgb.best_score_

-747.8885784028716

In [124]:
random_xgb.best_params_

{'subsample': 0.6,
 'n_estimators': 400,
 'min_child_weight': 16,
 'max_features': 30,
 'max_depth': 6,
 'learning_rate': 0.01,
 'colsample_bytree': 0.8}

In [125]:
random_xgb.best_estimator_

XGBRegressor(base_score=0.5, booster='gbtree', colsample_bylevel=1,
             colsample_bynode=1, colsample_bytree=0.8, gamma=0,
             importance_type='gain', learning_rate=0.01, max_delta_step=0,
             max_depth=6, max_features=30, min_child_weight=16, missing=None,
             n_estimators=400, n_jobs=-1, nthread=None, objective='reg:linear',
             random_state=3, reg_alpha=0, reg_lambda=1.2, scale_pos_weight=1,
             seed=None, silent=None, subsample=0.6, verbosity=1)

In [126]:
xgb = random_xgb.best_estimator_
xgb.fit(x_train,y_train)
pred = xgb.predict(x_test)

[12:54:16] WARNING: src/objective/regression_obj.cu:152: reg:linear is now deprecated in favor of reg:squarederror.


In [127]:
print('performance on test data')
print('Mean Absolute Error = ', metrics.median_absolute_error(y_test,pred))
print('Root Mean Square Error = ',my_rmse(y_test,pred))
print("Our model submission performance : ",1-(metrics.median_absolute_error(y_test,pred)/1600))

performance on test data
Mean Absolute Error =  498.7905494140625
Root Mean Square Error =  609433.9611021054
Our model submission performance :  0.6882559066162109


In [128]:
print("performance on train data")
pred_train = xgb.predict(x_train)
print('Mean Absolute Error = ', metrics.median_absolute_error(y_train,pred_train))
print('Root Mean Square Error = ',my_rmse(y_train,pred_train))
print("Our model submission performance : ",1-(metrics.median_absolute_error(y_train,pred_train)/1600))

performance on train data
Mean Absolute Error =  458.0763609375001
Root Mean Square Error =  486257.787829957
Our model submission performance :  0.7137022744140624


##### Final concluded parameters 

XGBRegressor(base_score=0.5, booster='gbtree', colsample_bylevel=1,
             colsample_bynode=1, colsample_bytree=0.8, gamma=0,
             importance_type='gain', learning_rate=0.01, max_delta_step=0,
             max_depth=6, max_features=30, min_child_weight=16, missing=None,
             n_estimators=400, n_jobs=-1, nthread=None, objective='reg:linear',
             random_state=3, reg_alpha=0, reg_lambda=1, scale_pos_weight=1,
             seed=None, silent=None, subsample=0.6, verbosity=1)

In [129]:
xgb = XGBRegressor(base_score=0.5, booster='gbtree', colsample_bylevel=1,
             colsample_bynode=1, colsample_bytree=0.8, gamma=0,
             importance_type='gain', learning_rate=0.01, max_delta_step=0,
             max_depth=6, max_features=30, min_child_weight=16, missing=None,
             n_estimators=400, n_jobs=-1, nthread=None, objective='reg:linear',
             random_state=3, reg_alpha=0, reg_lambda=1, scale_pos_weight=1,
             seed=None, silent=None, subsample=0.6, verbosity=1)

##### Building model on full training data and submission

In [130]:
x_training = training_data.drop('Counterfeit_Sales',axis='columns')
y_training = training_data['Counterfeit_Sales']

In [132]:
xgb.fit(x_training,y_training)
pred = xgb.predict(production_data)

[12:56:31] WARNING: src/objective/regression_obj.cu:152: reg:linear is now deprecated in favor of reg:squarederror.


In [133]:
production_medID = pd.read_csv(file_path + 'counterfeit_test.csv')

In [134]:
production_medID = production_medID['Medicine_ID']

In [135]:
submission_df = pd.DataFrame({"Medicine_ID": production_medID,
                 'Counterfeit_Sales': pred})

In [136]:
submission_df.to_csv('part2 project3.csv',index=False)

# Conclusion

### Linear Regression

Mean Absolute Error =  615.9795920660683

Root Mean Square Error =  641656.7346209903

Our model submission performance :  0.6150127549587072

### Decision Tree

Mean Absolute Error =  488.0314000000002

Root Mean Square Error =  625993.4594418913

Our model submission performance :  0.6949803749999999


### Random Forest

performance on test data

Mean Absolute Error =  491.96711024999945

Root Mean Square Error =  622689.0509111757

Our model submission performance :  0.6925205560937504

### Extra Trees

performance on test data

Mean Absolute Error =  473.6709262499995

Root Mean Square Error =  618246.3458160626

Our model submission performance :  0.7039556710937502

### Gradient Boost Machine

performance on test data

Mean Absolute Error =  481.02319571185

Root Mean Square Error =  604809.0510707761

Our model submission performance :  0.6993605026800938

### XG Boost

performance on test data

Mean Absolute Error =  494.82020566406254

Root Mean Square Error =  605384.4703138274

Our model submission performance :  0.690737371459961

